feat: provisionally support V2 libraries in LibraryContentBlock (randomized only)

kdmccormick opened the pull request:
Co-author credit to @connorhaugh and RaccoonGang.

kdmccormick committed:
feat!: remove LibrarySourcedBlock

Rather than implementing V2-library and static-library-reference support
in a new block, we will be enhancing the existing `LibraryContentBlock`
in-place.

Relevant ADR PR: https://github.com/openedx/edx-platform/pull/33231

kdmccormick committed:
feat: support V2 libraries in LibraryContentBlock (randomized only)

Co-Authored-By: Connor Haugh <chaugh@2u.com>
Co-Authored-By: Eugene Dyudyunov <evgen.dyudyunov@raccoongang.com>

kdmccormick commented on lms/envs/common.py:
```diff
@@ -3213,6 +3213,9 @@ def _make_locale_paths(settings):  # pylint: disable=missing-function-docstring
     'rules.apps.AutodiscoverRulesConfig',
     'bridgekeeper',
 
+    # management of user-triggered async tasks (course import/export, etc.)
+    'user_tasks',
+
```
Can we take this out, since import/export is still only CMS?

connorhaugh replied:
Tried to, haven't broken anything yet... (for context I did not add that code, and I don't know why it was there)

connorhaugh replied:
Ah -- the import content from blockstore task is being run from xmodule code, which complains about the LMS not having user tasks...

>  File "/edx/app/edxapp/edx-platform/openedx/core/djangoapps/course_groups/cohorts.py", line 21, in <module>
>     from lms.djangoapps.courseware import courses
>   File "/edx/app/edxapp/edx-platform/lms/djangoapps/courseware/courses.py", line 52, in <module>
>     from lms.djangoapps.courseware.block_render import get_block
>   File "/edx/app/edxapp/edx-platform/lms/djangoapps/courseware/block_render.py", line 49, in <module>
>     from xmodule.library_tools import LibraryToolsService
>   File "/edx/app/edxapp/edx-platform/xmodule/library_tools.py", line 19, in <module>
>     from user_tasks.models import UserTaskStatus
>   File "/edx/app/edxapp/venvs/edxapp/lib/python3.8/site-packages/user_tasks/models.py", line 31, in <module>
>     class UserTaskStatus(TimeStampedModel):
>   File "/edx/app/edxapp/venvs/edxapp/lib/python3.8/site-packages/django/db/models/base.py", line 113, in __new__
>     raise RuntimeError(
> RuntimeError: Model class user_tasks.models.UserTaskStatus doesn't declare an explicit app_label and isn't in an application in INSTALLED_APPS.

This is not something I am familiar with

connorhaugh replied:
maybe this is unique to my local environment? I'm running my tests from the cms shell.

kdmccormick replied:
Ah. That kinda makes sense. The task's top-level folder is part of what determines how the task is excuted. xmodule has never had celery tasks, so this is new territory. I'm a little fuzzy on the details but I think the platform is defaulting it to LMS workers.

It might be safer to put this task under cms/djangoapps/contentstore, along with all the other Studio tasks. If that'd cause an import cycle, then openedx/core/djangoapps/content_libraries is another place you could put it, although I think there is some decorator you'd need to add to make it run only in CMS.

connorhaugh replied:
Ok so I spent a solid day moving this tasks file to different locations. I think it should live with the openedx/core/djangoapps/content_libraries/tasks.py file, so the task to import content from a library into a course lives next to the task to import content from a library into a course. However, I can't seem to get these files to run in the cms, as it is not just a testing-level error. With that in mind, I'm going to keep 'user tasks' in the LMS, as there isn't a strong enough reason to NOT have it.

connorhaugh replied:
The only refactoring which is feasible is replacing the user tasks model with a bespoke one, which feels like a waste of energy.

connorhaugh replied:
https://github.com/openedx/edx-platform/blob/db459761d86ae1379344d6fcb8eda048674b8f80/.github/workflows/unit-test-shards.json#L157 THis might help me understand how tests are run & things are split up? Which doesn't solve this problem but might give some insight.

kdmccormick commented on cms/djangoapps/contentstore/tests/test_libraries.py:
```diff
@@ -993,24 +991,22 @@ def test_duplicated_version(self):
-        # But the block hasn't.
-        self.assertEqual(len(self.lc_block.children), 1)
-        self.assertEqual(self.problem_in_course.location, self.lc_block.children[0])
-        self.assertEqual(self.problem_in_course.display_name, self.original_display_name)
```
from @bradenmacdonald  : 
> This tests is no longer testing what it says in the docstring. I think it's important to preserve a test for that ("if a library content block is pinned to some library version and duplicated, the new block is pinned to the same version, and the children are NOT updated to the newest version")

kdmccormick commented on xmodule/assets/library_content/public/js/library_content_edit_helpers.js:
```diff
@@ -0,0 +1,54 @@
+    var libraryKey = libraries[selectedIndex].value;
+    var url = URI('/xblock')
+        .segment(editor.find('.xblock.xblock-studio_view.xblock-studio_view-library_content.xblock-initialized')
+        .data('usage-id'))
+        .segment('handler')
+        .segment('is_v2_library');
```
from @bradenmacdonald :  

> 
> **Nit**: this is an implicit dependency on a third-party `URI.js` library being present in the global scope.
> 
> Something like this would avoid the implicit dependency and is also easier to understand for people not familiar with `URI.js` (like me):
> ```js
> const usageId = editor.find('.xblock-studio_view-library_content').data('usage-id');
> const url = `/xblock/${usageId}/handler/is_v2_library`;
> ```
> I also think you can simplify the CSS selector, as shown.
> 
> If you do need a class to manipulate URLs, I would recommend using [URL](https://developer.mozilla.org/en-US/docs/Web/API/URL) and [URLSearchParams](https://developer.mozilla.org/en-US/docs/Web/API/URLSearchParams) which are web standard and don't increase the JS bundle size, instead of `URI.js`.

kdmccormick commented on xmodule/assets/library_content/public/js/library_content_edit_helpers.js:
```diff
@@ -0,0 +1,54 @@
+        success: function(data) {
+            var problemTypeSelect = editor.find("select[name='Problem Type']")
+                .parents("li.field.comp-setting-entry.metadata_entry");
+            data.is_v2 ? problemTypeSelect.hide() : problemTypeSelect.show();
```
from @bradenmacdonald :  
> It seems overkill to use a whole AJAX request to determine if this is a v2 library or not. The key formats are quite stable, so why not just check `const isV2 = usageId.startsWith("lb:");` (if true, it's v2; if false, it's v1).

kdmccormick replied:
@connorhaugh Brad mentioned page loading time as a release-blocking concern. Braden's suggestion here could help that.

connorhaugh replied:
Page load times in lib-authoring MFE. Not here. Load times for library-content-pick  are very quick.

kdmccormick commented on xmodule/library_content_block.py:
```diff
@@ -668,15 +701,23 @@ def source_library_values(self):
+
+    def post_editor_saved(self):
+        """
+        If xblock has been edited, refresh_children automatically.
```
Open discussion: https://github.com/openedx/edx-platform/pull/33057#discussion_r1307945409

connorhaugh replied:
So in summary: the post_editor_saved event is fired after the editor is saved, the editor_saved event fires when the author presses the save button. I am now sure of this, and can see the benefits of adding this lifecycle method, but am open to suggestions as to how we want to refactor it out, if anyone can think of any.

kdmccormick commented on xmodule/library_tools.py:
```diff
@@ -137,15 +170,14 @@ def can_use_library_content(self, block):
+        The children will be given new block_ids.
+
+        NOTE: V1 libraies blocks definition ID should be the
+        exact same definition ID used in the copy block.
```
I'm not sure what this is saying; we should either clarify or remove it.

The definition IDs will be different between the source blockstore blocks and the copied modulestore blocks, as blockstore and modulestore use different key formats. However, the definition IDs between the library_content children and the modulestore result of `import_from_blockstore` may be the same. I'm not sure yet; I'll look into this.

kdmccormick commented on xmodule/library_tools.py:
```diff
@@ -27,39 +36,63 @@ def normalize_key_for_search(library_key):
         self.user_id = user_id
 
-    def _get_library(self, library_key):
+    def _get_library(self, library_key, is_v2_lib):
```
from @bradenmacdonald  :  

> 
> Nit: it's often ambiguous whether parameters like `library_key` here are accepting a library key or a string. I would suggest adding the explicit type hint `library_key: LearningContextKey` declaration.
> 
> And it seems odd to accept `is_v2_lib` as a parameter when it can be trivially deduced from the key - 
> 
> Instead of `if is_v2_lib:` use `if isinstance(lib_key, LibraryLocatorV2):`

kdmccormick replied:
Connor addressed this

kdmccormick commented on xmodule/library_tools.py:
```diff
@@ -155,48 +187,66 @@ def update_children(self, dest_block, user_perms=None, version=None):
+
+        if not is_v2_lib:
+            if user_perms and not user_perms.can_read(library_key):
+                raise PermissionDenied()
```
As @bradenmacdonald points out, the second half of this block (everything under `if not is_v2_lib`) is redundant with the first half of the block.

kdmccormick commented on xmodule/tasks.py:
```diff
@@ -0,0 +1,319 @@
+"""
+This module contains tasks for asynchronous execution of library xblocks.
```
```suggestion
This module contains tasks for asynchronous updates of library xblocks.
```

kdmccormick commented on xmodule/tasks.py:
```diff
@@ -0,0 +1,319 @@
+    return capa_type in descriptor.problem_types
+
+
+def _get_library(store, library_key, is_v2_lib):
```
Dupe of _get_library in library_tools?

connorhaugh replied:
Yup, but can't be having circular imports.

kdmccormick replied:
@connorhaugh Ok, the logic is pretty simple, should be fine.

+1 to Braden's suggestion of replacing `is_v2_lib` with `isinstance(library_key, LibraryLocatorV2)`, in both instances of the function.

kdmccormick commented on xmodule/tasks.py:
```diff
@@ -0,0 +1,319 @@
+
+def _import_block(store, user_id, source_block, dest_parent_key):
+    """
+    Recursively import a blockstore block and its children. See import_from_blockstore above.
```
```suggestion
    Recursively import a blockstore block and its children. See import_from_blockstore.
```

It's not above any more; let's make the docstring agnostic of where in the file the function is.

kdmccormick commented on xmodule/tasks.py:
```diff
@@ -0,0 +1,319 @@
+def _filter_child(store, usage_key, capa_type):
+    """
+    Filters children by CAPA problem type, if configured
+    """
```
```suggestion
def _filter_child(store, usage_key, capa_type):
    """
    Return whether this block is a problem of type `capa_type`.
    """
```

Docstring is wrong. Here's a more accurate one.

kdmccormick reviewed (commented):
Carrying over all outstanding comments from https://github.com/openedx/edx-platform/pull/33057

kdmccormick committed:
Merge branch 'master' into kdmccormick/library-content-v2

Commit:
fix: user tasks out of lms, move tasks.py

Commit:
Merge commit 'refs/pull/33263/head' of github.com:openedx/edx-platform into kdmccormick/library-content-v2

Commit:
fix: lint

Commit:
fix: lint

Commit:
fix: lint fix

Commit:
fix: move codeowner attribute to layer

kdmccormick committed:
fix: remove codeowner attribute from update_children_task

...per comment in cms/djangoapps/contentstore/tasks.py.

This fixes an exception that was raised in the background whenever
the update_children_task was started:

     Traceback (most recent call last):
       File "/openedx/venv/lib/python3.8/site-packages/celery/utils/dispatch/signal.py", line 276, in send
         response = receiver(signal=self, sender=sender, **named)
       File "/openedx/venv/lib/python3.8/site-packages/user_tasks/signals.py", line 215, in start_user_task
         sender.status.start()
       File "/openedx/venv/lib/python3.8/site-packages/user_tasks/tasks.py", line 84, in status
         name = self.generate_name(arguments_dict)
       File "/openedx/edx-platform/openedx/core/djangoapps/content_libraries/tasks.py", line 274, in generate_name
         key = arguments_dict['dest_block_key']
     KeyError: 'dest_block_key'

kdmccormick committed:
build: put back set_code_owner_attribute but as a regular function call

kdmccormick committed:
feat: add cms-only assertions to content_libraries tasks

kdmccormick committed:
Merge branch 'master' into kdmccormick/library-content-v2

kdmccormick committed:
style: fix indentation

kdmccormick committed:
fix: add missing imports for _assert_cms

kdmccormick committed:
docs: remove outdated is_v2_lib from function docstring

kdmccormick committed:
Merge remote-tracking branch 'upstream/master' into kdmccormick/library-content-v2

kdmccormick committed:
test: run ./xmodule/ tests with CMS settings

Just like ./common/ and ./openedx/, the unit tests in ./xmodule/
validate behavior in both LMS and CMS. In order to fully test
./xmodule/, we need to run its tests in both contexts.

Also in this commit:

* refactor: Rename the shards to be clear whether they're
            running under LMS or CMS.
* docs: correct comments regarding conditions under which
        codejail's test_cant_do_something_forbidden is skipped.

* test: update a unit test which was using the now-deleted
        library_sourced block to use library_content block instead.

kdmccormick committed:
test: run ./xmodule/ LibraryContentTests only in CMS

kdmccormick committed:
test: run ./xmodule/ LibraryContentTests only in CMS

kdmccormick committed:
Revert "test: run ./xmodule/ LibraryContentTests only in CMS"

This reverts commit f1370d0b33674925ed0240567f3fb1ebfcd2a814.

kdmccormick committed:
Merge remote-tracking branch 'upstream/master' into kdmccormick/library-content-v2

kdmccormick committed:
Merge branch 'kdmccormick/test-xmodule-with-cms' into kdmccormick/library-content-v2

kdmccormick committed:
style: remove duplicate imports fomr test_lib_tools

kdmccormick committed:
test: fix patching in test_list_available_libraries

kdmccormick committed:
test: attempt at stricter enforcement of library tools in cms only

kdmccormick committed:
Merge branch 'master' into kdmccormick/library-content-v2

kdmccormick committed:
temp: fix: put library_tools back in LMS, at least for now

kdmccormick committed:
test: fix skip_unless_[lms|cms]

kdmccormick committed:
test: fix library_content completion test w/ settings hack

kdmccormick committed:
Merge remote-tracking branch 'upstream/master' into kdmccormick/library-content-v2

kdmccormick committed:
test: update test patch for new libraries tasks location

kdmccormick committed:
test: use string mock to satisfy importlinter

kdmccormick committed:
Merge remote-tracking branch 'upstream/master' into kdmccormick/library-content-v2

kdmccormick commented on openedx/core/lib/__init__.py:
```diff
@@ -0,0 +1,40 @@
+        )
+
+
+def ensure_cms(message: str = "This code may only be called by CMS, but it was called by LMS"):
```
@connorhaugh This module, and a few usages of this ensure_cms function, are the only significant new thing I added to the PR. Everything else is just tweaks and unit test / linting fixes.

connorhaugh reviewed (approved):
Thumb counts insofar as I wrote a lot of this code and think it is good!

kdmccormick committed:
Merge remote-tracking branch 'upstream/master' into kdmccormick/library-content-v2

kdmccormick commented:
Right now, I'm manually testing this in `tutor local`, which actually runs the celery tasks asynchronously.

kdmccormick committed:
fix: getting status of library_content's update_children task

kdmccormick committed:
fix: getting status of library_content's update_children task

kdmccormick committed:
Merge remote-tracking branch 'upstream/master' into kdmccormick/library-content-v2

kdmccormick committed:
fix: squash: _are_children_updating utiltity func

kdmccormick committed:
Revert "fix: squash: _are_children_updating utiltity func"

This reverts commit 535fb52ffd554f38adbdf02d49f124714883dcd8.

kdmccormick committed:
Revert "fix: getting status of library_content's update_children task"

This reverts commit 0cf735fcab69b0c011e1991bc7aa5ea113dc1399.

kdmccormick committed:
Merge remote-tracking branch 'upstream/master' into kdmccormick/library-content-v2

kdmccormick committed:
Merge remote-tracking branch 'upstream/master' into kdmccormick/library-content-v2

connorhaugh commented on openedx/core/djangoapps/content_libraries/tasks.py:
```diff
@@ -39,3 +78,281 @@ def on_progress(block_key, block_num, block_count, exception=None):
+        return [key for key in library.children if _filter_child(store, key, capa_type)]
+
+
+def import_from_blockstore(user_id, store, dest_block, blockstore_block_ids):
```
Note: we have this `import_from_blockstore` (and its dependancies) code both here and in `library_tools.py` . Currently thinking about which place is the correct place to put this code, but just don't want to have redundant or unused code so it is worthwhile to remove one.

i'll put a comment where this code is duplicated.

kdmccormick replied:
Good catch @connorhaugh .

My thinking is that `tasks.py` is the right place for this code rather than the library_tools service class. That's because `import_from_blockstore` isn't really a "service" that we want to provide to XBlocks. The service we want to provide is `update_children`, and `import_from_blockstore` is just an implementation detail to achieve that.

Also, from a practical standpoint, keeping this function in `library_tools.py` would prevent an import cycle: `library_tools:LibraryTools.update_children` -> `tasks.update_children_task` -> `library_tools:LibraryToolsService:import_from_blockstore`

kdmccormick replied:
Done. Latest commit should fix tests.

FYI, two other changes I made:
* `update_children` is now `trigger_update_children_task`.
* there was a test that specifically tested `_import_from_blockstore` (which is now a private function in `tasks.py`), so I updated it to test the same functionality but through `trigger_update_children_task`.

connorhaugh commented on xmodule/library_tools.py:
```diff
@@ -155,40 +187,53 @@ def update_children(self, dest_block, user_perms=None, version=None):
+            return v2_libs
+        return v1_libs + v2_libs
 
     def import_from_blockstore(self, dest_block, blockstore_block_ids):
```
This is is duplicate code in `contentstore/tasks.py`

kdmccormick committed:
Merge branch 'master' into kdmccormick/library-content-v2

kdmccormick committed:
refactor: delete unused library_tools code; rename update_children

kdmccormick committed:
test: update unit tests for previous refactor commit

kdmccormick commented on xmodule/library_content_block.py:
```diff
@@ -534,7 +572,7 @@ def studio_post_duplicate(self, store, source_block):
-        self.tools.update_children(self, user_perms, version=self.source_library_version)
+        self.tools.trigger_update_children_task(self, user_perms, version=self.source_library_version)
 
         self._copy_overrides(store, user_id, source_block, self)
```
@connorhaugh I think there's a race-condition bug here. Because `update_children` now happens in a Celery task, `_copy_overrides` is likely to be invoked *before* the children are imported into the duplicate lib content block, which seems to prevent those new children from properly getting their overrides from the original lib content block.

Reproduction:
* in the unit editor in Studio, add a library content block, pointing at any type of library
* click "View All" to see the library content block's children
* override the titles of one of the child blocks. go back to the unit editor.
* on the library content block, hit the dots, hit "duplicate"
* click into "View All" of the new duplicate lib content block.

Expected behavior (confirmed on master):
* the child's overriden title is also overriden in the duplicate lib content block.

Actual behavior (confirmed on this branch):
* the child has the original title from the source library.

I'm thinking about potential fixes currently. I don't know whether we need to make `_copy_overrides` happen after `update_children_task` finishes, or if we  can change `_copy_overrides` to work if even if children aren't imported yet.

kdmccormick replied:
I think what I will try to do is to make a new `duplicate_library_content_block_task` Celery task, which calls out to update_children first, and then calls out to _copy_overrides next. All the code could live in `tasks.py` so we don't have circular imports.

connorhaugh replied:
It is taking me a while and some exploration to understand why this "update children, then copy overrides" operation is necessary.
If, for original block A and duplicate block B, we simply set B's children to be a list of copies of A's children, then we would be sorted, right?
The only reason I can understand why this this problem was solved using update_children is to provide the following expected behavior:

1. Block A has 3 children. You change block 1's name and delete block 2.
2. You copy Block A to make block B.
3. You expect block 2 to be in block B's children.

The other possible reasons this was solved this way are coding choices: we want to maintain as we possible paths to do a thing, these are the tools I already have, etc.

I know your solution (waiting for update_children_task to finish) will work, and if that is easy (which it may not be) that will work, but this is maybe worth exploring.

connorhaugh replied:
> I think what I will try to do is to make a new duplicate_library_content_block_task Celery task, which calls out to update_children first, and then calls out to _copy_overrides next. All the code could live in tasks.py so we don't have circular imports


I agree with this approach

kdmccormick replied:
@connorhaugh You know, I honestly don't know why we're importing-then-copying-overrides rather than just copying the blocks!

kdmccormick replied:
@connorhaugh Here's why: https://2u-internal.atlassian.net/browse/SOL-549 (I found it via [this PR](https://github.com/openedx/edx-platform/pull/7482), which added the special-handling of LCB duplication).

I'm not going to say I fully understand it, but at least there's clearly a reason.

kdmccormick committed:
Merge remote-tracking branch 'upstream/master' into kdmccormick/library-content-v2

open-craft-grove commented:
Sandbox deployment successful.

Sandbox LMS is available at [pr-33263-139931.sandboxes.opencraft.hosting](https://pr-33263-139931.sandboxes.opencraft.hosting)
Sandbox Studio is available at [studio.pr-33263-139931.sandboxes.opencraft.hosting](https://studio.pr-33263-139931.sandboxes.opencraft.hosting)

gabor-boros commented:
/grove sandbox update

open-craft-grove commented:
Sandbox update request received. Deployment will start soon.

open-craft-grove commented:
Sandbox deployment started.

open-craft-grove commented:
Sandbox deployment successful.

Sandbox LMS is available at [pr-33263-139931.sandboxes.opencraft.hosting](https://pr-33263-139931.sandboxes.opencraft.hosting)
Sandbox Studio is available at [studio.pr-33263-139931.sandboxes.opencraft.hosting](https://studio.pr-33263-139931.sandboxes.opencraft.hosting)

kdmccormick committed:
Merge remote-tracking branch 'upstream/master' into kdmccormick/library-content-v2

open-craft-grove commented:
Sandbox update request received. Deployment will start soon.

open-craft-grove commented:
Sandbox deployment started.

open-craft-grove commented:
Sandbox deployment successful.

Sandbox LMS is available at [pr-33263-139931.sandboxes.opencraft.hosting](https://pr-33263-139931.sandboxes.opencraft.hosting)
Sandbox Studio is available at [studio.pr-33263-139931.sandboxes.opencraft.hosting](https://studio.pr-33263-139931.sandboxes.opencraft.hosting)

kdmccormick committed:
fix: duplicate library_content children asynchronously (#33652)

open-craft-grove commented:
Sandbox update request received. Deployment will start soon.

open-craft-grove commented:
Sandbox deployment started.

connorhaugh commented on xmodule/library_content_block.py:
```diff
@@ -527,16 +547,13 @@ def studio_post_duplicate(self, store, source_block):
-
-        self._copy_overrides(store, user_id, source_block, self)
+        self.tools.trigger_duplicate_children(
+            user_id=user_id, user_perms=user_perms, source_block=source_block, dest_block=self
```
extraneous user_id param, as user_id is acquired from the library tools object itself.

kdmccormick replied:
👍🏻

kdmccormick committed:
fix: bad user_id kwarg in call to trigger_duplicate_children

kdmccormick committed:
Merge remote-tracking branch 'upstream/master' into kdmccormick/library-content-v2

kdmccormick committed:
test: reorder library_content test_duplicate_version assertions for better debugging

open-craft-grove commented:
Sandbox update request received. Deployment will start soon.

open-craft-grove commented:
Sandbox deployment successful.

Sandbox LMS is available at [pr-33263-139931.sandboxes.opencraft.hosting](https://pr-33263-139931.sandboxes.opencraft.hosting)
Sandbox Studio is available at [studio.pr-33263-139931.sandboxes.opencraft.hosting](https://studio.pr-33263-139931.sandboxes.opencraft.hosting)

kdmccormick committed:
fix: various fixes relating to library version awareness (WIP)

open-craft-grove commented:
Sandbox update request received. Deployment will start soon.

open-craft-grove commented:
Sandbox deployment started.

open-craft-grove commented:
Sandbox deployment successful.

Sandbox LMS is available at [pr-33263-139931.sandboxes.opencraft.hosting](https://pr-33263-139931.sandboxes.opencraft.hosting)
Sandbox Studio is available at [studio.pr-33263-139931.sandboxes.opencraft.hosting](https://studio.pr-33263-139931.sandboxes.opencraft.hosting)

kdmccormick committed:
fix: fail better when library_tools is missing

kdmccormick committed:
refactor: rename to be clearer when we're upgrading lib versions vs just syncing

kdmccormick committed:
Merge remote-tracking branch 'upstream/master' into kdmccormick/library-content-v2

open-craft-grove commented:
Sandbox update request received. Deployment will start soon.

kdmccormick committed:
test: change 'refresh' to 'reselect' in library_content test name

"refreshing" was ambiguous, possibly meaning either:

a. syncing the lc block's children with the library, or
b. reselecting the particular set of lc block children that learner
sees.

open-craft-grove commented:
Sandbox deployment started.

open-craft-grove commented:
Sandbox update request received. Deployment will start soon.

open-craft-grove commented:
Sandbox deployment started.

kdmccormick committed:
fix: differentiate upgrade_and_sync http handler from its internal method

the handler checks permissions, the internal method doesn't.

this is necessary to fix tests. it also makes the internal method
consistent with sync_from_library, which doesn't check perms either.

open-craft-grove commented:
Sandbox update request received. Deployment will start soon.

open-craft-grove commented:
Sandbox deployment started.

kdmccormick commented:
@connorhaugh  While fixing unit tests, I found that we were calling refresh_children all over the place in tests where it didn't make any sense to do so. E.g., when testing changes to max_count , we shouldn't rely on refresh_children being called--simply calling selected_children() should do the right thing.

I think this is at least partially due to the fact that refresh_children is an ambiguous name. It definitely has been confusing me, at least. Just based on the name, it's not clear whether it (a) updates to the latest library, (b) pulls blocks in from the library, or (c) just re-selects blocks for a single learner. In reality, the function can do either (a+b) or (b) depending on its args, and then (c) is a downstream consequence of either of those.

So, I've turned it into sync_from_library (b), with upgrade_to_latest (a) as an optional flag. The XBlock HTTP handler is renamed from refresh_children to upgrade_and_sync. LMK if you disagree with any of those names.

I realize this will be minor PITA for your rebase.

kdmccormick committed:
fix: make get_latest_library_version always return str

kdmccormick committed:
fix: use get_tools() consistently. also, turn upgrade_to_latest into flag

open-craft-grove commented:
Sandbox update request received. Deployment will start soon.

open-craft-grove commented:
Sandbox deployment successful.

Sandbox LMS is available at [pr-33263-139931.sandboxes.opencraft.hosting](https://pr-33263-139931.sandboxes.opencraft.hosting)
Sandbox Studio is available at [studio.pr-33263-139931.sandboxes.opencraft.hosting](https://studio.pr-33263-139931.sandboxes.opencraft.hosting)

connorhaugh commented:
>  I found that we were calling refresh_children all over the place in tests where it didn't make any sense to do so. E.g., when testing changes to max_count , we shouldn't rely on refresh_children being called--simply calling selected_children() should do the right thing. 

Yeah I have seen it done haphazardly. I agree with the new naming scheme and the optional flag. Those two changes shouldn't collide too heavily with work that I have done, but we shall see. While I'm wary that perfect is enemy of good enough in this case, this will hopefully future proof this code in a worthwhile way.

kdmccormick committed:
test(libraries): only sync when necessary, and only upgrade when necessary

kdmccormick committed:
fix: don't stringify lib version when it's None

kdmccormick commented:
@connorhaugh I believe this is code complete and that the next build will be green. I'm gonna take a walk and then do some manual testing. Feel free to review, although I'll let you know once I've manually tested in case you want to wait.

kdmccormick commented on cms/djangoapps/contentstore/tests/test_libraries.py:
```diff
@@ -466,17 +467,23 @@ def test_refresh_fails_for_unknown_library(self):
         self.assertEqual(resp.status_code, 200)
+        lc_block = modulestore().get_item(lc_block.location)
+        assert lc_block.source_library_key == LibraryLocator(org="NOT", library="FOUND")
+        assert lc_block.source_library_version is None
```
Layering in some additional assertions which I found helpful when debugging this test earlier. Nothing really new.

kdmccormick commented on cms/djangoapps/contentstore/xblock_storage_handlers/view_handlers.py:
```diff
@@ -387,12 +398,11 @@ def _save_xblock(  # lint-amnesty, pylint: disable=too-many-statements
     fields=None,
     summary_configuration_enabled=None,
-):
+):  # lint-amnesty, pylint: disable=too-many-statements
```
```suggestion
):  # pylint: disable=too-many-statements
```
nit: we shouldn't be adding lint-amnesty to new pylints

kdmccormick commented on openedx/core/djangoapps/content_libraries/api.py:
```diff
@@ -668,7 +671,7 @@ def get_library_blocks(library_key, text_search=None, block_types=None):
     # If indexing is disabled, or connection to elastic failed
     if metadata is None:
         metadata = []
-        ref = ContentLibrary.objects.get_by_key(library_key)
+        ref = ContentLibrary.objects.get_by_key(library_key)  # type: ignore[attr-defined]
```
mypy is getting confused by custom object managers, oh well

kdmccormick commented on openedx/core/djangoapps/content_libraries/api.py:
```diff
@@ -1165,6 +1168,77 @@ def revert_changes(library_key):
+#  https://github.com/openedx/edx-platform/issues/32457)
+# ======================================================
+
+def get_v1_or_v2_library(
+    library_id: str | LibraryLocatorV1 | LibraryLocatorV2,
+    version: str | int | None,
+) -> LibraryRootV1 | ContentLibraryMetadata | None:
```
This function is a little crazy, but I figure it's better to handle all the edge cases in one place than to scatter them throughout.

kdmccormick commented on openedx/core/djangoapps/content_libraries/api.py:
```diff
@@ -1165,6 +1168,77 @@ def revert_changes(library_key):
+            v2_version = None
+        try:
+            library = get_library(library_key)
+            if v2_version is not None and library.version != v2_version:
+                raise NotImplementedError(
+                    f"Tried to load version {v2_version} of blockstore-based library {library_key}. "
+                    f"Currently, only the latest version ({library.version}) may be loaded. "
+                    "This is a known issue. "
+                    "It will be fixed before the production release of blockstore-based (V2) content libraries. "
+                )
```
Here's that fail-if-we-need-a-v2-library-other-than-latest temporary workaround we talked about. I'll follow up on this next.

kdmccormick commented on xmodule/library_content_block.py:
```diff
@@ -466,12 +485,14 @@ def non_editable_metadata_fields(self):
         ])
         return non_editable_fields
 
-    @lazy
-    def tools(self):
+    def get_tools(self, to_read_library_content: bool = False) -> 'LibraryToolsService':
```
Because loading library-tools is something that can fail, it seemed wrong to put it into a property.

I also add a new flag: `to_library_content`. When passed true, This function also asserts that `can_use_library_content` is True. This is the case for any operation which needs to read library content blocks, like syncing or duplication.

kdmccormick commented on xmodule/library_content_block.py:
```diff
@@ -485,40 +506,76 @@ def get_user_id(self):
             user_id = None
         return user_id
 
+    def _validate_sync_permissions(self):
+        """
+        Raises PermissionDenied() if we can't confirm that user has write on this block and read on source library.
+
+        If source library isn't set, then that's OK.
+        """
+        if not (user_perms := self.runtime.service(self, 'studio_user_permissions')):
+            raise PermissionDenied("Access cannot be validated in the current runtime.")
+        if not user_perms.can_write(self.scope_ids.usage_id.context_key):
+            raise PermissionDenied(f"Cannot write to block at {self.scope_ids.usage_id}")
+        if self.source_library_key:
+            if not user_perms.can_read(self.source_library_key):
+                raise PermissionDenied(f"Cannot read library at {self.source_library_key}")
```
We need these permission checks both for duplication and for upgrade_and_sync, so I pulled them out to here.

kdmccormick commented on xmodule/library_content_block.py:
```diff
@@ -666,17 +716,34 @@ def source_library_values(self):
+        """
+        If xblock has been edited, upgrade library & sync automatically.
+
+        TODO: This shouldn't happen if only post-library settings like max_count are changed.
```
@connorhaugh This is what we talked about in Slack.

kdmccormick commented on xmodule/library_tools.py:
```diff
@@ -27,43 +22,34 @@ def normalize_key_for_search(library_key):
         self.user_id = user_id
 
-    def _get_library(self, library_key):
+    def get_latest_library_version(self, lib_key) -> str | None:
```
get_library_version is now get_latest_library_version. Same functionality, just a more specific name.

open-craft-grove commented:
Sandbox update request received. Deployment will start soon.

kdmccormick commented on xmodule/tests/test_library_content.py:
```diff
@@ -53,13 +57,25 @@ def setUp(self):
         )
+        self.lc_block.runtime._services.update({'library_tools': self.tools})  # pylint: disable=protected-access
+
+    def _sync_lc_block_from_library(self, upgrade_to_latest=False):
+        """
+        Save the lc_block, then sync its children with the library, and then re-load it.
+
+        We must re-load it because the syncing happens in a Celery task, so that original self.lc_block instance will
+        not have changes manifested on it, but the re-loaded instance will.
+        """
+        self.store.update_item(self.lc_block, self.user_id)
+        self.lc_block.sync_from_library(upgrade_to_latest=upgrade_to_latest)
+        self.lc_block = self.store.get_item(self.lc_block.location)
```
the save/sync/reload pattern was need in a ton of test cases, so I pulled up to here.

kdmccormick commented on xmodule/tests/test_library_content.py:
```diff
@@ -71,16 +87,59 @@ def get_block(descriptor):
+        )
+        assert isinstance(library.source_library_key, expected_locator_type)
+
+    def test_initial_sync_from_library(self):
+        """
+        Test that a lc block starts without children, but is correctly populated upon first sync.
+        """
+        source_library_key = self.library.location.library_key
+
+        # Normally the children get added when the "source_libraries" setting
+        # is updated, but the way we do it through a factory doesn't do that.
+        assert self.lc_block.source_library_key == source_library_key
+        assert self.lc_block.source_library_version is None
+        assert len(self.lc_block.children) == 0
+
+        # Update the LibraryContent block's children:
+        self._sync_lc_block_from_library()
+
+        # Check that all blocks from the library are now children of the block:
+        assert self.lc_block.source_library_key == source_library_key  # Unchanged
+        assert self.lc_block.source_library_version == self.tools.get_latest_library_version(source_library_key)
+        assert len(self.lc_block.children) == len(self.lib_blocks)
```
A new test I added after struggling with bugs around the "uninitialized" edge case. This case is relevant when an LC block's library has been set, but the celery task to sync its children hasn't finished yet.

kdmccormick commented on xmodule/tests/test_library_content.py:
```diff
@@ -177,6 +236,10 @@ class LibraryContentBlockTestMixin:
 
     problem_type_lookup = {}
 
+    def setUp(self):
+        super().setUp()
+        self._sync_lc_block_from_library()
```
Every test case in this class needed the sync, so I put it in setUp.

kdmccormick commented on xmodule/tests/test_library_content.py:
```diff
@@ -196,26 +259,10 @@ def _create_capa_problems(self):
             block = self.make_block("problem", self.library, data=self._get_capa_problem_type_xml(*problem_type))
             self.problem_type_lookup[block.location] = problem_type
 
-    def test_lib_content_block(self):
-        """
-        Test that blocks from a library are copied and added as children
-        """
-        # Check that the LibraryContent block has no children initially
-        # Normally the children get added when the "source_libraries" setting
-        # is updated, but the way we do it through a factory doesn't do that.
-        assert len(self.lc_block.children) == 0
-        # Update the LibraryContent block:
-        self.lc_block.refresh_children()
-        self.lc_block = self.store.get_item(self.lc_block.location)
-        # Check that all blocks from the library are now children of the block:
-        assert len(self.lc_block.children) == len(self.lib_blocks)
-
```
Is now part of `test_initial_sync_from_library`.

open-craft-grove commented:
Sandbox deployment started.

kdmccormick commented on xmodule/tests/test_library_content.py:
```diff
@@ -234,121 +281,147 @@ def test_validation_of_course_libraries(self):
+        self._sync_lc_block_from_library()
         assert self.lc_block.validate()
 
+        # But updating the library will cause it to fail again as out-of-date:
+        self._add_problems_to_library()
+        result = self.lc_block.validate()
+        assert not result
+        assert result.summary
+        assert StudioValidationMessage.WARNING == result.summary.type
+        assert 'out of date' in result.summary.text
+
+        # And a regular sync will not fix that:
+        self._sync_lc_block_from_library()
+        result = self.lc_block.validate()
+        assert not result
+        assert result.summary
+        assert StudioValidationMessage.WARNING == result.summary.type
+        assert 'out of date' in result.summary.text
+
+        # But a upgrade_to_latest sync will:
+        self._sync_lc_block_from_library(upgrade_to_latest=True)
+        assert self.lc_block.validate()
+
+    def _assert_has_only_N_matching_problems(self, result, n):
+        assert result.summary
+        assert StudioValidationMessage.WARNING == result.summary.type
+        assert f'only {n} matching problem' in result.summary.text
+
```
Some new assertion, to check the cases where a library has been updated and then a non-upgrade sync occurs (i.e., an import happens), and _then_ an upgrade-sync occurs.

kdmccormick commented on xmodule/tests/test_library_content.py:
```diff
@@ -234,121 +281,147 @@ def test_validation_of_course_libraries(self):
+        self._sync_lc_block_from_library(upgrade_to_latest=True)
         self.lc_block.max_count = 1
-        self._create_capa_problems()
-        self.lc_block.refresh_children()
         assert self.lc_block.validate()
```
I strengthened this test a bit: by setting max_count and then validating _without_ syncing from library, we confirm that reselection of a learner's blocks isn't contingent on a library sync.

kdmccormick reviewed (commented):
Notes for reviewers

kdmccormick committed:
fix: lib version type annoations; also remove extraneous update_items

open-craft-grove commented:
Sandbox update request received. Deployment will start soon.

open-craft-grove commented:
Sandbox deployment successful.

Sandbox LMS is available at [pr-33263-139931.sandboxes.opencraft.hosting](https://pr-33263-139931.sandboxes.opencraft.hosting)
Sandbox Studio is available at [studio.pr-33263-139931.sandboxes.opencraft.hosting](https://studio.pr-33263-139931.sandboxes.opencraft.hosting)

open-craft-grove commented:
Sandbox deployment started.

open-craft-grove commented:
Sandbox deployment failed.

Please check the settings and requirements and retry deployment.

connorhaugh commented:
I found a bug with it not pulling the correct block version when duplicating. I am on commit 2a5c85e091f0a48d4e82be9b40a116b952e4519b (which seems to be head, just confirming). LMK if you can't replicate this on tutor.

Here's what I did to manually test this PR

1. Checkout, requirements, update static assets.
2. Created a unit
3. Created a V1 lib, `v1lib`, and a v2lib, `v2lib`
4. Created a lib block in the unit, associated the v1lib with it.
5. set max count to be -1
6. View the problems imported from the lib.
7. Publish the unit, view it in the LMS, see one problem (at random)
8. Make a change to v1lib
9. View the block's contents
10. Duplicate the block in the unit.
11.View the duplicate's contents **(Bug: 🐛 IT pulls in the latest version of the v1 and v2 lib when doing this atm with no unit-local overrides)**
12. Go back to the original block's children, make an edit to one of them in a field you edited in the library previously
13. duplicate that original block again.
14. See that you pull in the changes you just made in the original block
15. publish the unit, make sure that all blocks are shown
16. Do 4-15 for v2 blocks (making sure to hit publish when you make changes to the lib this time)

kdmccormick commented:
/grove sandbox create

open-craft-grove commented:
Sandbox update request received. Deployment will start soon.

open-craft-grove commented:
Sandbox deployment successful.

Sandbox LMS is available at [pr-33263-139931.sandboxes.opencraft.hosting](https://pr-33263-139931.sandboxes.opencraft.hosting)
Sandbox Studio is available at [studio.pr-33263-139931.sandboxes.opencraft.hosting](https://studio.pr-33263-139931.sandboxes.opencraft.hosting)

kdmccormick commented:
/grove sandbox destroy

open-craft-grove commented:
Sandbox update request received. Deployment will start soon.

open-craft-grove commented:
Sandbox deploy request received. Deployment will start soon.

open-craft-grove commented:
Sandbox deployment started.

open-craft-grove commented:
Sandbox deployment failed.

Please check the settings and requirements and retry deployment.

open-craft-grove commented:
Sandbox deployment started.

open-craft-grove commented:
Sandbox deployment failed.

Please check the settings and requirements and retry deployment.

kdmccormick committed:
test: fix setup for import/export tests

kdmccormick committed:
Merge remote-tracking branch 'upstream/master' into kdmccormick/library-content-v2

open-craft-grove commented:
Sandbox update request received. Deployment will start soon.

open-craft-grove commented:
Sandbox deployment started.

open-craft-grove commented:
Sandbox deployment failed.

Please check the settings and requirements and retry deployment.

kdmccormick committed:
fix: post_editor_saved should only apply when source lib or capa type changes

kdmccormick committed:
Merge remote-tracking branch 'upstream/master' into kdmccormick/library-content-v2

kdmccormick committed:
fix: correct post_editor saved (wip - test still failing

open-craft-grove commented:
Sandbox update request received. Deployment will start soon.

open-craft-grove commented:
Sandbox deployment started.

open-craft-grove commented:
Sandbox deployment failed.

Please check the settings and requirements and retry deployment.

kdmccormick commented:
Alright @connorhaugh , I believe I've gotten this branch to handle duplication of V1 LC blocks correctly. I used [this unit edX Stage](https://studio.stage.edx.org/container/block-v1:edx+jrr_103+H2_2022+type@vertical+block@67c8b485325d401296953c7636f3df4a) as my reference point for what the "correct behavior is". My testing process was:

* Create a library with two text components A and B.
* Create a course unit with a randomized content block (RCB) pointing at that lib and Count==-1.
* Back in the lib, edit both the title and content of A and B, which on the backend should trigger new versions ("New Ver") of the lib to exist. Should have:
  ![image](https://github.com/openedx/edx-platform/assets/3628148/294c4bde-a3eb-49f9-afe4-a1586d692d6a)
* Back in the course, within RCB, override both the title and content of A. The component should be "out of date". Should be:
  ![image](https://github.com/openedx/edx-platform/assets/3628148/0fa63460-0216-44e7-809b-180237362fc1)
* Duplicate RCB (we'll call it DRCB). We expect that the library version is not updated, and that the title override (but NOT the content override) is preserved:
  ![image](https://github.com/openedx/edx-platform/assets/3628148/ec7ed449-1c69-44a2-a3a3-91a7bf74963b)
* Note: it seems wrong that DRCB's version of A doesn't have overridden content, but that's an artifact of how LC components handle duplication, and I confirmed on stage, so :shrug: I added it to [my running list of follow-ups](https://github.com/openedx/edx-platform/issues/33640) as a bug we might want to fix.

connorhaugh commented:
> Alright @connorhaugh , I believe I've gotten this branch to handle duplication of V1 LC blocks correctly. I used [this unit edX Stage](https://studio.stage.edx.org/container/block-v1:edx+jrr_103+H2_2022+type@vertical+block@67c8b485325d401296953c7636f3df4a) as my reference point for what the "correct behavior is". My testing process was:

Yeah I can confirm that this testing process worked. I did it myself. 

v2 is not fully functional to duplicate because we decided that would be follow-on work (we need to update the content libraries v2 api itself to do that). It silently failed with "Tried to load version 1 of blockstore-based library lib:edx:ABV2LibTest. Currently, only the latest version (2) may be loaded. This is a known issue. It will be fixed before the production release of blockstore-based (V2) content libraries. "

Doing some more manual testing and code reading on this myself to make sure nothing else regressed.

kdmccormick committed:
fix: return 400 (not 500) when upgrade_and_sync called with bad lib key

kdmccormick committed:
test: consolidate & fix test cases around nonexistent source libraries

kdmccormick committed:
Merge remote-tracking branch 'upstream/master' into kdmccormick/library-content-v2

kdmccormick committed:
fix: typos in last two commits

open-craft-grove commented:
Sandbox update request received. Deployment will start soon.

kdmccormick commented:
I pushed up a fix for that failing unit test. Now I'm writing a unit test for LCB duplication.

> v2 is not fully functional to duplicate because we decided that would be follow-on work (we need to update the content libraries v2 api itself to do that). It silently failed with "Tried to load version 1 of blockstore-based library lib:edx:ABV2LibTest. Currently, only the latest version (2) may be loaded. This is a known issue. It will be fixed before the production release of blockstore-based (V2) content libraries. "

Yup, right. That makes sense.

> Doing some more manual testing and code reading on this myself to make sure nothing else regressed.

Great, same.

open-craft-grove commented:
Sandbox deployment started.

kdmccormick commented on cms/djangoapps/contentstore/tests/test_libraries.py:
```diff
@@ -458,26 +495,6 @@ def test_refreshes_children_if_capa_type_change(self):
         html_block = modulestore().get_item(lc_block.children[0])
         self.assertEqual(html_block.display_name, name2)
 
-    def test_refresh_fails_for_unknown_library(self):
-        """ Tests that refresh children fails if unknown library is configured """
-        # Create a course:
-        with modulestore().default_store(ModuleStoreEnum.Type.split):
-            course = CourseFactory.create()
-
-        # Add a LibraryContent block to the course:
-        lc_block = self._add_library_content_block(course, self.lib_key)
-        lc_block = self._refresh_children(lc_block)
-        self.assertEqual(len(lc_block.children), 0)
-
-        # Now, change the block settings to have an invalid library key:
-        resp = self._update_block(
-            lc_block.location,
-            {"source_library_id": "library-v1:NOT+FOUND"},
-        )
-        self.assertEqual(resp.status_code, 200)
-        with self.assertRaises(ValueError):
-            self._refresh_children(lc_block, status_code_expected=400)
-
```
This test, which was failing, heavily overlapped with test_change_after_sync. In fixing it, I found it easiest to smash them together into one big test: test_switch_to_unknown_source_library_preserves_settings.

kdmccormick commented on cms/djangoapps/contentstore/tests/test_libraries.py:
```diff
@@ -326,10 +326,11 @@ def test_block_with_children(self):
         self.assertEqual(course_child_block.data, data_value)
         self.assertEqual(course_child_block.display_name, name_value)
 
-    def test_change_after_first_sync(self):
```
Renamed this test, and beefed it up with assertions from test_refresh_fails_for_unknown_library.

kdmccormick commented on cms/djangoapps/contentstore/tests/test_libraries.py:
```diff
@@ -348,31 +349,58 @@ def test_change_after_first_sync(self):
+        self.assertEqual(len(lc_block.children), 1)
+        self.assertEqual(modulestore().get_item(lc_block.children[0]).data, data_value)
+
+    def test_sync_if_source_library_changed(self):
+        """
+        Tests that children are automatically synced with new lib if source library id is changed.
+        """
```
This block was originally:
```python
    def test_refreshes_children_if_libraries_change(self):
        """ Tests that children are automatically refreshed if libraries list changes """
```
which was just not an accurate description of what it tests.

kdmccormick commented on cms/djangoapps/contentstore/tests/test_libraries.py:
```diff
@@ -389,25 +417,34 @@ def test_refreshes_children_if_libraries_change(self):
+        html_block_1 = modulestore().get_item(lc_block.children[0])
+        self.assertEqual(html_block_1.data, data1)
 
-        # Now, change the block settings to have an invalid library key:
```
Inaccurate comment

open-craft-grove commented:
Sandbox deployment failed.

Please check the settings and requirements and retry deployment.

kdmccormick committed:
test: test_duplicate_library_content_block

kdmccormick committed:
Merge remote-tracking branch 'upstream/master' into kdmccormick/library-content-v2

open-craft-grove commented:
Sandbox update request received. Deployment will start soon.

kdmccormick committed:
fix: import shouldn't fail if lib doesn't exist

open-craft-grove commented:
Sandbox update request received. Deployment will start soon.

open-craft-grove commented:
Sandbox deployment failed.

Please check the settings and requirements and retry deployment.

open-craft-grove commented:
Sandbox deployment started.

open-craft-grove commented:
Sandbox deployment failed.

Please check the settings and requirements and retry deployment.

kdmccormick commented:
@connorhaugh Unit tests are all set now. The only failing check is a random pylint thing that I can fix Monday.

I'll be manually testing this more Monday morning. Assuming the recent commits look good to you and we don't find bugs, let's merge in the afternoon?

connorhaugh commented:
@kdmccormick I'm also giving it one final once over.

kdmccormick committed:
style: ignore pylint violation

kdmccormick committed:
refactor: undo accidental test case rename

kdmccormick committed:
Merge remote-tracking branch 'upstream/master' into kdmccormick/library-content-v2

open-craft-grove commented:
Sandbox deployment started.

open-craft-grove commented:
Sandbox update request received. Deployment will start soon.

open-craft-grove commented:
Sandbox deployment failed.

Please check the settings and requirements and retry deployment.

open-craft-grove commented:
Sandbox update request received. Deployment will start soon.

open-craft-grove commented:
Sandbox update request received. Deployment will start soon.

open-craft-grove commented:
Sandbox update request received. Deployment will start soon.
